In [ ]:
with open("the-verdict.txt", "r" , encoding="utf-8") as f:
    raw_text=f.read()

#prints all the characters
print("Total number of characters:" , len(raw_text))

#prints first 100 characters
print(raw_text[:99])


Total number of characters: 20479
I HAD always thought Jack Gisburn rather a cheap genius--though a good fellow enough--so it was no 


<div class="alert alert-block alert-success">

Our goal is to tokenize this 20,479-character short story into individual words and special
characters that we can then turn into embeddings for LLM training  </div>

<div class="alert alert-block alert-warning">

Note that it's common to process millions of articles and hundreds of thousands of
books -- many gigabytes of text -- when working with LLMs. However, for educational
purposes, it's sufficient to work with smaller text samples like a single book to
illustrate the main ideas behind the text processing steps and to make it possible to
run it in reasonable time on consumer hardware. </div>

In [ ]:
import re 

text = "hello world, this is a test"

# splits the white spaces
result = re.split(r'(\s)',text)

print(result)

['hello', ' ', 'world,', ' ', 'this', ' ', 'is', ' ', 'test']


In [ ]:
text = "hello world, this is a test"

result= re.split(r'([,.]|\s)',text)
print(result)

['hello', ' ', 'world', ',', '', ' ', 'this', ' ', 'is', ' ', 'a', ' ', 'test']


In [9]:
text = "Hello, world. Is this-- a test?"

result=re.split(r'([.,:;!?()\']|--|\s)',text)
result=[item.strip() for item in result if item.strip()]
print(result)

['Hello', ',', 'world', '.', 'Is', 'this', '--', 'a', 'test', '?']


In [15]:
preprocessed = re.split(r'([,.;?\'()_!":]|--|\s)',raw_text)
preprocessed = [item.strip() for item in preprocessed if item.strip()]
print(preprocessed[:30])

['I', 'HAD', 'always', 'thought', 'Jack', 'Gisburn', 'rather', 'a', 'cheap', 'genius', '--', 'though', 'a', 'good', 'fellow', 'enough', '--', 'so', 'it', 'was', 'no', 'great', 'surprise', 'to', 'me', 'to', 'hear', 'that', ',', 'in']


In [16]:
print(len(preprocessed))

4690


In [18]:
all_words=sorted(set(preprocessed))
vocab_size=len(all_words)
print(vocab_size)

1130


In [19]:
vocab = {token:integer for integer,token in enumerate(all_words)}

In [21]:
for i,item in enumerate(vocab.items()):
    print(item)
    if i>=20:
        break

('!', 0)
('"', 1)
("'", 2)
('(', 3)
(')', 4)
(',', 5)
('--', 6)
('.', 7)
(':', 8)
(';', 9)
('?', 10)
('A', 11)
('Ah', 12)
('Among', 13)
('And', 14)
('Are', 15)
('Arrt', 16)
('As', 17)
('At', 18)
('Be', 19)
('Begin', 20)


In [29]:
class SimpleTokenizerV1:
    def __init__(self,vocab):
        self.str_to_int = vocab
        self.int_to_str = {i:s for s,i in vocab.items()}

    def encode(self,text):
        preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)', text)
        preprocessed =[item.strip() for item in preprocessed if item.strip()]

        ids=[self.str_to_int[i] for i in preprocessed]
        return ids
    
    def decode(self,ids):
        text=" ".join([self.int_to_str[s] for s in ids])
        # Replace spaces before the specified punctuations
        text = re.sub(r'\s+([,.?!"()\'])', r'\1', text)
        return text

        

In [30]:
tokenizer = SimpleTokenizerV1(vocab)

text = """"It's the last he painted, you know," 
           Mrs. Gisburn said with pardonable pride."""
ids = tokenizer.encode(text)

print(ids)

text=tokenizer.decode(ids)

print(text)

[1, 56, 2, 850, 988, 602, 533, 746, 5, 1126, 596, 5, 1, 67, 7, 38, 851, 1108, 754, 793, 7]
" It' s the last he painted, you know," Mrs. Gisburn said with pardonable pride.


In [ ]:
#it gives an error when a word is not present inside 
# text = "Hello, do you like tea?"
# print(tokenizer.encode(text))

In [40]:
all_tokens = sorted(list(set(preprocessed)))

all_tokens.extend(["<|endoftext|>","<|unk|>"])

vocab={token:integer for integer, token in enumerate(all_tokens)}

In [34]:
print(len(vocab))

1132


In [36]:
for i, item in enumerate(list(vocab.items())[-5:]):
    print(item)

('younger', 1127)
('your', 1128)
('yourself', 1129)
('<|endoftext|', 1130)
('<|unk|>', 1131)


In [41]:
class SimpleTokenizerV2:
    def __init__(self,vocab):
        self.str_to_int = vocab
        self.int_to_str = {i:s for s,i in vocab.items()}

    def encode(self,text):
        preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)', text)
        preprocessed=[item.strip() for item in preprocessed if item.strip()]
        
        preprocessed = [
            item if item in self.str_to_int
            else "<|unk|>" for item in preprocessed
        ]
        ids = [self.str_to_int[i] for i in preprocessed]
        return ids
    
    def decode(self,ids):
        text=" ".join([self.int_to_str[s] for s in ids])
        # Replace spaces before the specified punctuations
        text = re.sub(r'\s+([,.?!"()\'])', r'\1', text)
        return text
        

In [42]:
tokenizer = SimpleTokenizerV2(vocab)

text1 = "Hello, do you like tea?"
text2 = "In the sunlit terraces of the palace."

text = " <|endoftext|> ".join((text1, text2))

print(text)

Hello, do you like tea? <|endoftext|> In the sunlit terraces of the palace.


In [43]:
en=tokenizer.encode(text)

print(en)

print(tokenizer.decode(en))

[1131, 5, 355, 1126, 628, 975, 10, 1130, 55, 988, 956, 984, 722, 988, 1131, 7]
<|unk|>, do you like tea? <|endoftext|> In the sunlit terraces of the <|unk|>.


In [45]:
! pip3 install tiktoken

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 994.0/994.0 kB 12.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2/2 [tiktoken]

[notice] A new release of pip is available: 25.1.1 -> 25.3
[notice] To update, run: pip install --upgrade pip


In [46]:
import importlib
import tiktoken

print("tiktoken version", importlib.metadata.version("tiktoken"))


tiktoken version 0.12.0


In [54]:
tokenizer = tiktoken.get_encoding("gpt2")

# its similar to SimpleTokenizerV2 we created

In [55]:
text = (
    "Hello, do you like tea? <|endoftext|> In the sunlit terraces"
     "of someunknownPlace."
)

integers = tokenizer.encode(text, allowed_special={"<|endoftext|>"})

print(integers)

[15496, 11, 466, 345, 588, 8887, 30, 220, 50256, 554, 262, 4252, 18250, 8812, 2114, 1659, 617, 34680, 27271, 13]


In [56]:
strings = tokenizer.decode(integers)

print(strings)

Hello, do you like tea? <|endoftext|> In the sunlit terracesof someunknownPlace.


In [57]:
integers = tokenizer.encode("Akwirw ier")

print(integers)

strings = tokenizer.decode(integers)
print(strings)

[33901, 86, 343, 86, 220, 959]
Akwirw ier


# CREATING INPUT-TARGET PAIRS

In [66]:
with open("the-verdict.txt","r") as f:
    raw_text=f.read()
enc_text = tokenizer.encode(raw_text)
print(len(enc_text))

5145


In [67]:
#removing the first 50 tokens from the dataset
enc_sample = enc_text[50:]


In [68]:
context_size=4 #length of the input

#the context size means that the model is trained to look at a sequence of 4 words or tokens
# to predict the next word of the sequence
#the input x is the first 4 tokens [1,2,3,4] and the target y is the next 4 tokens [2,3,4,5]

x= enc_sample[:context_size]
y = enc_sample[1:context_size+1]

print(f"x: {x}")
print(f"y:      {y}")



x: [290, 4920, 2241, 287]
y:      [4920, 2241, 287, 257]


In [69]:
for i in range(1,context_size+1):
    context=enc_sample[:i]
    desired=enc_sample[i]

    print(context , "----->",desired)

[290] -----> 4920
[290, 4920] -----> 2241
[290, 4920, 2241] -----> 287
[290, 4920, 2241, 287] -----> 257


In [70]:
for i in range(1,context_size+1):
    context=enc_sample[:i]
    desired=enc_sample[i]

    print(tokenizer.decode(context) , "---->" , tokenizer.decode([desired]))

 and ---->  established
 and established ---->  himself
 and established himself ---->  in
 and established himself in ---->  a


##  IMPLEMENTING A DATA LOADER

In [91]:
from torch.utils.data import Dataset, DataLoader
import torch
# import numpy


class GPTDatasetV1(Dataset):
    def __init__(self,txt,tokenizer,max_length,stride):
        self.input_ids = []
        self.target_ids = []

        #tokenize the entire text
        token_ids=tokenizer.encode( txt ,allowed_special={"<|endoftext|>"})

        #Use a sliding window to chunk the book into overlapping sequences of max_length(context_size)
        for i in range(0,len(token_ids)-max_length , stride):
            input_chunk = token_ids[i:i+max_length]
            target_chunk = token_ids[i+1:i+1+max_length]
            
            self.input_ids.append(torch.tensor(input_chunk))
            self.target_ids.append(torch.tensor(target_chunk))
    def __len__(self):
        return len(self.input_ids)
    
    def __getitem__(self, idx):
        return self.input_ids[idx] ,self.target_ids[idx]

        

In [92]:
def create_dataloader_v1(txt,batch_size=4,max_length=256,
                        stride=128,shuffle=True,drop_last=True,
                        num_workers=0):
    
    #Initialize the tokenizer
    tokenizer=tiktoken.get_encoding("gpt2")

    #create Dataset
    dataset=GPTDatasetV1(txt,tokenizer,max_length,stride)

    #create dataloader
    dataloader= DataLoader(
        dataset,
        batch_size=batch_size,
        shuffle=shuffle,
        drop_last=drop_last,
        num_workers=num_workers
    )

    return dataloader

In [93]:
with open("the-verdict.txt","r") as f:
    raw_text=f.read()

In [94]:
import torch
print("PyTorch version:", torch.__version__)

dataloader=create_dataloader_v1(raw_text,batch_size=1,max_length=4,stride=1,shuffle=False)

data_iter = iter(dataloader)
first_batch=next(data_iter)
print(first_batch)



PyTorch version: 2.9.1
[tensor([[  40,  367, 2885, 1464]]), tensor([[ 367, 2885, 1464, 1807]])]


In [95]:
next_batch = next(data_iter)
print(next_batch)

#stride is 1 here as inputs just shifted by 1


[tensor([[ 367, 2885, 1464, 1807]]), tensor([[2885, 1464, 1807, 3619]])]


In [ ]:
data_loader= create_dataloader_v1(raw_text,batch_size=8, max_length=4, stride=4, shuffle=False)



data_iter = iter(dataloader)
inputs, targets = next(data_iter)
print("Inputs:\n", inputs)
print("\nTargets:\n", targets)

Inputs:
 tensor([[  40,  367, 2885, 1464]])

Targets:
 tensor([[ 367, 2885, 1464, 1807]])
